File: SandBank.API/Services/AuthService.cs
using Microsoft.EntityFrameworkCore;
using Microsoft.AspNetCore.Identity;
using SandBank.Data;
using SandBank.DTOs;
using SandBank.Models;
using BCrypt.Net;

namespace SandBank.Services;

public class AuthService : IAuthService
{
    private readonly ApplicationDbContext _context;
    private readonly ITokenService _tokenService;

    public AuthService(ApplicationDbContext context, ITokenService tokenService)
    {
        _context = context;
        _tokenService = tokenService;
    }

    public async Task<AuthResponseDto> Register(RegisterDto registerDto)
    {
        if (await _context.Users.AnyAsync(u => u.Email == registerDto.Email))
        {
            throw new Exception("User already exists");
        }

        var passwordHash = BCrypt.Net.BCrypt.HashPassword(registerDto.Password);

        var newUser = new User
        {
            Username = registerDto.Username,
            Email = registerDto.Email,
            PasswordHash = passwordHash,
        };

        _context.Users.Add(newUser);
        await _context.SaveChangesAsync();

        var token = _tokenService.GenerateToken(newUser);

        return new AuthResponseDto
        {
            Token = token,
            Username = newUser.Username,
            Email = newUser.Email,
            ExpiresAt = DateTime.UtcNow.AddMinutes(60)
        };
    }

    public async Task<AuthResponseDto> Login(LoginDto loginDto)
    {
        // Find user by email
        var user = await _context.Users.FirstOrDefaultAsync(u => u.Email == loginDto.Email);

        if (user == null)
        {
            throw new Exception("User not found");
        }

        bool isPasswordValid = BCrypt.Net.BCrypt.Verify(loginDto.Password, user.PasswordHash);

        if (!isPasswordValid)
        {
            throw new Exception("Password not valid");
        }

        var token = _tokenService.GenerateToken(user);

        return new AuthResponseDto
        {
            Token = token,
            Username = user.Username,
            Email = user.Email,
            ExpiresAt = DateTime.UtcNow.AddMinutes(60)
        };
    }
}

File: SandBank.API/services/auth_service.py
from sqlalchemy.orm import Session
import bcrypt
from datetime import datetime, timedelta, timezone
from models.user import User
from schemas.auth import RegisterSchema, LoginSchema, AuthResponseSchema
from services.token_service import generate_token

def hash_password(password: str) -> str:
    return bcrypt.hashpw(password.encode("utf-8"), bcrypt.gensalt()).decode("utf-8")

def verify_password(plain: str, hashed: str) -> bool:
    return bcrypt.checkpw(plain.encode("utf-8"), hashed.encode("utf-8"))

def register(db: Session, data: RegisterSchema) -> AuthResponseSchema:
    existing = db.query(User).filter(User.email == data.email).first()
    if existing:
        raise Exception("User already exists")

    new_user = User(
        username=data.username,
        email=data.email,
        password_hash=hash_password(data.password)
    )

    db.add(new_user)
    db.commit()
    db.refresh(new_user)

    token = generate_token(new_user)

    return AuthResponseSchema(
        id=new_user.id, 
        token=token,
        username=new_user.username,
        email=new_user.email,
        role=new_user.role,
        expires_at=datetime.now(timezone.utc) + timedelta(minutes=60)
    )

def login(db: Session, data: LoginSchema) -> AuthResponseSchema:
    user = db.query(User).filter(User.email == data.email).first()
    if not user:
        raise Exception("User not found")

    if not verify_password(data.password, user.password_hash):
        raise Exception("Password not valid")

    token = generate_token(user)

    return AuthResponseSchema(
        id=user.id,
        token=token,
        username=user.username,
        email=user.email,
        role=user.role,
        expires_at=datetime.now(timezone.utc) + timedelta(minutes=60)
    )